main/src/library/Iterator.flix

```diff
@@ -20,6 +20,269 @@ pub enum Iterator[a] {
+    ///
+    /// If `iter` contains multiple mappings with the same key, `toMap` does not
+    /// make any guarantees about which mapping will be in the resulting map.
+    ///
+    /// Consumes the entire iterator.
+    ///
+    pub def toMap(iter: Iterator[(a, b)]): Map[a, b] & Impure with Order[a] =
+        toMapHelper(iter, Map.empty())
+
+    ///
+    /// Helper function for `toMap`
+    ///
+    def toMapHelper(iter: Iterator[(a, b)], m: Map[a, b]): Map[a, b] & Impure with Order[a] =
+        let Iterator(done, next) = iter;
+        if (done())
+            m
+        else
+            let (a, b) = next();
+            toMapHelper(iter, Map.insert(a, b, m))
+
+    ///
+    /// Returns the iterator `iter` as a set.
+    ///
+    /// Consumes the entire iterator.
+    ///
+    pub def toSet(iter: Iterator[a]): Set[a] & Impure with Order[a] =
+        toSetHelper(iter, Set.empty())
+
+    ///
+    /// Helper function for `toSet`
+    ///
+    def toSetHelper(iter: Iterator[a], xs: Set[a]): Set[a] & Impure with Order[a] =
+        let Iterator(done, next) = iter;
+        if (done())
+            xs
+        else
+            let x = next();
+            toSetHelper(iter, Set.insert(x, xs))
+
+    ///
+    /// Returns `iter` with every occurrence of `x` replaced by `y`.
+    ///
+    /// Does *not* consume any elements of the iterator.
+    ///
+    pub def replace(x: a, y: a, iter: Iterator[a]): Iterator[a] with Eq[a] =
+        let Iterator(done, next) = iter;
+        let n = () -> {
+            let el = next();
+            if (el == x) y else el
+        };
+        Iterator(done, n)
+
+    ///
+    /// Returns `true` if and only if at least one element in `iter` satisfies the predicate `f`.
+    ///
+    /// Returns `false` if `iter` is empty.
+    ///
+    /// The function `f` must be pure.
+    ///
+    /// Consumes elements of the iterator until an element that satisfies predicate `f` is found.
+    ///
+    pub def exists(f: a -> Bool, iter: Iterator[a]): Bool & Impure =
+        let Iterator(done, next) = iter;
+        if (done())
+            false
+        else
+            f(next()) or exists(f, iter)
+
+    ///
+    /// Returns `true` if and only if all elements in `iter` satisfy the predicate `f`.
+    ///
+    /// Returns `true` if `iter` is empty.
+    ///
+    /// The function `f` must be pure.
+    ///
+    /// Consumes the entire iterator.
+    ///
+    pub def forall(f: a -> Bool, iter: Iterator[a]): Bool & Impure =
+        let Iterator(done, next) = iter;
+        if (done())
+            true
+        else
+            f(next()) and forall(f, iter)
+
+    ///
+    /// Returns `iter` without the longest prefix that satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    /// Consumes elements of the iterator until the first element that does *not* satisfy predicate `f`.
+    ///
+    pub def dropWhile(f: a -> Bool, iter: Iterator[a]): Iterator[a] & Impure =
```

Comment:
Add a TODO that this likely needs to be overloaded based on the purity/impurity of f.